addons/mail/static/src/components/activity/activity.scss

```diff
@@ -79,108 +46,22 @@
 // Style
 // ------------------------------------------------------------------
 
-$o-mail-activity-default-color: gray('300') !default;
-$o-mail-activity-overdue-color: o-text-color('danger') !default;
-$o-mail-activity-planned-color: o-text-color('success') !default;
-$o-mail-activity-today-color: o-text-color('warning') !default;
-
-
-.o_Activity_deadlineDateText {
-   &.o-default {
-       color: $o-mail-activity-default-color;
-   }
-
-   &.o-overdue {
-       color: $o-mail-activity-overdue-color;
-   }
-
-   &.o-planned {
-       color: $o-mail-activity-planned-color;
-   }
-
-   &.o-today {
-    color: $o-mail-activity-today-color;
-   }
-}
-
-.o_Activity_details {
-    color: gray('500');
-}
-
-.o_Activity_detailsCreatorAvatar {
-    margin-inline-start: map-get($spacers, 2);
-}
-
-.o_Activity_detailsUserAvatar {
-    @extend .rounded-circle;
-}
-
-.o_Activity_dueDateText {
-    font-weight: bolder;
-
-   &.o-default {
-       color: $o-mail-activity-default-color;
-   }
-
-   &.o-overdue {
-       color: $o-mail-activity-overdue-color;
-   }
-
-   &.o-planned {
-       color: $o-mail-activity-planned-color;
-   }
-
-   &.o-today {
-    color: $o-mail-activity-today-color;
-   }
-}
-
-/* Needed specifity to counter default bootstrap style */
-a:not([href]):not([tabindex]).o_Activity_detailsButton {
-    background: transparent;
-    opacity: 0.5;
-    color: gray('500');
-
-    &:hover {
-        opacity: 1;
-        color: gray('600');
-    }
-}
-
-.o_Activity_detailsCreator {
-    font-weight: bold;
+.o_Activity_detailsButton {
+    @include o-hover-text-color($default-color: $o-main-color-muted);
 }
 
 .o_Activity_iconContainer {
-    color: $white;
-    border-color: $white;
-    border-radius: 100%;
-    border-style: solid;
-}
-
-.o_Activity_sidebar {
-    font-size: smaller;
-}
-
-.o_Activity_summary {
-    font-weight: bolder;
-    color: gray('900');
+    box-shadow: 0 0 0 2px currentColor;
 }
 
 .o_Activity_toolButton {
-    opacity: 0.5;
-    color: gray('500');
+    @include o-hover-text-color();
 
-    &:hover {
-        opacity: 1;
-        color: gray('600');
+    .o_Activity:not(:hover) & {
```

Comment:
> isn't "not hover" just standard style?

not-hover is the "exception" that use the rule `opacity: .6;`, since by default is `opacity: 1;`

Otherwise I'll have to define the rule twice, eg:
```scss
.o_Activity & {
        opacity: .6;
 }
.o_Activity:hover & {
        opacity: 1;
 }
```

Reply:
Well except `o-hover-text-color` also puts the style on `&:hover, &:focus, &.focus` and you don't negate the others, so it's not working as intended.

It's much easier to understand when defining both rules like the existing code (just change value from 0.5 to 0.6 if you need) and like in your example.

Also what is this extra `&` exactly? Same comment here as elsewhere, don't use specificity and put class where they are needed.

Reply:
> Well except o-hover-text-color also puts the style on &:hover, &:focus, &.focus and you don't negate the others, so it's not working as intended.

I think you missed the point. We don't care if `o_Activity` is :focus, we only care when the user interact hovering it ;)

> It's much easier to understand when defining both rules like the existing code

Yes but it's 3 lines more + an unnecessary override. I prefer to add 3 lines of notes if it's not clear.

> Also what is this extra & exactly?

`&` is not "extra" ,  it's the sass parent selector.
https://sass-lang.com/documentation/style-rules/parent-selector

Reply:
We definitely don't want to ever use parent selector, it's just magically adding style on elements from random context.
Same as usual, if we need a style on the parent in that case, then the parent should have a class for that style on itself when necessary.

Reply:
> We definitely don't want to ever use parent selector, it's just magically adding style on elements from random context.

You do realise that it's equivalent to write  `.o_Activity:not(:hover) .o_Activity_toolButton {}`, right ?
There is nothing "magic"... it's just taking advantage of the preprocessor for what it is supposed to do, simplify the code.

Anyways... it's my understanding we could spend ages debating on a single line, but we don't have the time ^^
I'll try to stay as close as possible to what you want.

Could you just share with me the guidelines your team uses?
What you suggest is not in the official ones  https://github.com/odoo/odoo/wiki/SCSS-coding-guidelines

Reply:
> > We definitely don't want to ever use parent selector, it's just magically adding style on elements from random context.
> 
> You do realise that it's equivalent to write `.o_Activity:not(:hover) .o_Activity_toolButton {}`, right ? There is nothing "magic"... it's just taking advantage of the preprocessor for what it is supposed to do, simplify the code.

The underlying problem with this rule is that it deals with higher CSS specificity than `.o_Activity_toolButton` local rules. At first this doesn't look like a big deal, but actually that's precisely what makes most of the SCSS a nightmare to maintain when CSS specificity is spread all over the style of the whole application.

The more maintainable approach consists of this: all style-rules target immediately the classname of elements to style, and contextual styling come only from "modifier" classnames on these elements. That way, CSS specificity is always the same for all static styles (= 1) and for all contextual styles (= 2). This design incites all styling on an element being pretty local in the code, which is very good for code maintenance.

Unfortunately, doing so is quite cumbersome right now, as you need to implement modifier classnames by hand and integrate some built-in feature of CSS in JS/XML like `:hover`... We actually did this in many places and we are quite happy to have them instead of dealing with specificity in CSS everywhere. So it's easier to maintain, but unfortunately it requires more effort to make it work initially.

TBH, it's not urgent to do the proper thing right now. At least, the problem with `.o_Activity:not(:hover) .o_Activity_toolButton` is only limited to elements from the `Activity` component, which is better than cross-component styling for instance! But this clearly worse than the approach with modifier classnames, because potentially anything in the SCSS file can have styling effect on any part of the component rather than just the style on element.

Let's say discuss code is around 80 SCSS files and each average 100 LOCs, and all styles that target a given classname average to about 10 LOCs:
1) Dealing with different CSS specificities in all discuss SCSS files, I would need to know all combination of CSS styling of all files at all time, which is 80 * 100 LOCs, so **8000 LOCs**.
2) Dealing with different CSS specificities in each discuss SCSS file, I would need to know all combination of CSS styling of single file at a given time, so **100 LOCs**
3) The approach of minimising CSS specificity wth immediate target classname and modifier classnames, so **10 LOCs**.

Just quantitatively, 8000 > 100 > 10.
Please note that dealing with CSS specificity in cases 1 and 2 make branching for evaluating styles to be applied on some elements. These are much much harder to work with. IMO I'd say the effort to maintain CSS is exponential so even between 100 > 10 the maintenance cost is quite significant.

Thank you for attending my TED Talk 😅

Reply:
To complement what Alex just posted.

We follow more or less BEM guidelines (Block Element Modifier) slightly adapted to how things are usually named in Odoo (often to simplify a bit). I think it would be interesting that you read on the subject if you haven't heard of it before.

Each rule should only target one specific class, not children, not parent, not siblings, and there should be no nested classes either.

The goal is when we look at any DOM element, we want to know exactly which style applies to it by looking only at its own class attribute.
Because not doing it this way means any line of CSS anywhere in the code could have an impact on any possible element, which is completely unmaintainable, like existing CSS files.


In this particular case, if we need to apply style on `.o_Activity_toolButton` conditionally, then the element with `.o_Activity_toolButton` should receive the conditional class (example: `.o-activityFocused`) with a `t-aff` or `t-attf` when it is necessary, which means we need to store the focus state in our modelization so that it gets rendered correctly with the class on the element from the template. Of course I'm available to help you with that part if you need, I can give you the code necessary to add the class in the DOM, so that you can use it for style.

Reply:
Very happy to see that we share a lot of opinions.
I don't think it's a surprise: this shows we are able to take a step back with the tools we use and pinpoint precisely some of their issues.

> Modularity & Reusability

Fully agree that currently the code of Discuss component is not reusable, which is clearly not efficient.

Actually all of this comes from our initial experience with Odoo Widgets / OWL components: code architecture with them strongly motivates to design the code around UI elements in DOM, and not really care with the state of application. For very small and simple codebase, this is decent while giving fast results. But when your modelling becomes complex due to functional scope, this scales very poorly.

In Discuss, we work with models first and then components. Components is about UI so if 2 parts of UI share some styles and it's not a coincidence, this means these UI elements share a same model. Our priority is fixing our mistakes with Widget/Component of the past and properly use a model-oriented architecture. Reusability of component is a much lower priority at the moment, as things are still relatively maintainable even with code duplication.
Also the huge work on model-oriented development will fix component reusability, because we will have proper model like `EntriesList`, so we will define components are proper UI things rather than very specific components.

So I fully agree with you on that part, and it takes a lot of time to improve that properly.

> Combined selectors

If the 2 styles are related and it's not by coincidence, this means there is an underlying concept that is shared.
In our model-oriented way of developing the Discuss codebase, there is a model that is shared between the 2 elements.

Instead of guessing what is the concept, prefer to actually introduce it. If you do so, then you don't actually need combined selectors.

> Split selectors in layout/style

We are not sure to keep this split. We just followed the idea is some other SCSS files and like some advantages of this split ("layout" is macro-view of style where  changes can affect the whole screen, while "style" is very local and safe to change).

It's a bit cumbersome to make the split, even though there's clearly a gain in the split.
IMO this is not a big deal as the split layout/style is arbitrary split in a single SCSS file, and if we follow the methodology of minimising CSS specificity with BEM, then it's a non-issue w.r.t code maintainability.

Just my opinion, but I think layout/style split makes sense, but unfortunately everything is mixed in one technology (CSS), and this declarative language a bit of a mess...